fixed extends parsing #32

JayHuLBL opened the pull request:
This closes #32.

The `extends` clause (from https://github.com/ibpsa/modelica-ibpsa/blob/master/IBPSA/Fluid/FixedResistances/HydraulicDiameter.mo):
```
  extends IBPSA.Fluid.FixedResistances.PressureDrop(
    final deltaM =  eta_default*dh/4*Modelica.Constants.pi*ReC/m_flow_nominal_pos,
    final dp_nominal=fac*dpStraightPipe_nominal);
```

can be parsed as:
```
"extends_clause": {
                    "name": "IBPSA.Fluid.FixedResistances.PressureDrop",
                    "class_modification": {
                      "modifications": [
                        {
                          "prefix": "final",
                          "name": "deltaM",
                          "value": "eta_default*dh*4*Modelica.Constants.pi*ReC*m_flow_nominal_pos"
                        },
                        {
                          "prefix": "final",
                          "name": "dp_nominal",
                          "value": "fac*dpStraightPipe_nominal"
                        }
                      ]
                    }
                  }
```

mwetter commented:
@JayHuLBL : The change looks good, but as it adds a new functionality, you will also need to add one or several test cases in `test/FromModelica` These tests are to illustrate the parsing, and also to make sure your code does not inadvertently gets changed and the result is suddenly different.

JayHuLBL commented:
@mwetter Thanks for the remind. I will add the tests.

mwetter commented:
Can you please also add a simple example of a modelica model that has two `extends` statements,
and one modelica file that has an `extends` statement with a `redeclare` in the `extends` statement.

JayHuLBL commented:
@mwetter 
- [x] updated from Antlr4.3 to 4.7,  used a new Antlr function to further optimize the code
- [x] added two more test cases for `extends_clause` to parse:
```
  extends Buildings.Controls.OBC.CDL.Continuous.LimPID(
    k=2,
    Ti=2);
  extends Buildings.Controls.OBC.CDL.Logical.TrueHoldWithReset(
    redeclare Buildings.Controls.OBC.CDL.Logical.TrueDelay onDelay(
      final delayTime=duration),
    final duration = 300);
```

mwetter commented on lib/jsonquery.js:
```diff
@@ -465,6 +475,50 @@ function parsecomponentDeclaration (declaration) {
+
+function parseExtendsAssignments (classModification) {
+  if (classModification === undefined) {
+    return {}
```
This should return `null` and not `{}`

mwetter commented on lib/jsonquery.js:
```diff
@@ -465,6 +475,50 @@ function parsecomponentDeclaration (declaration) {
 }
 
+function parseextendsClause (extendsClause) {
+  var ret = {
```
Fix capitalization to `parseExtendsClause`

mwetter commented on lib/jsonquery.js:
```diff
@@ -465,6 +475,50 @@ function parsecomponentDeclaration (declaration) {
+  var assignments = []
+  const mod = classModification.modifications
+
+  for (var k = 0; k < mod.length; k++) {
```
Use `var assignments = classModification.modifications.map(element => ....)` rather than a loop

mwetter commented on test/FromModelica/ExtendsClause-1.html:
```diff
@@ -0,0 +1,251 @@
+      Buildings.Controls.OBC.CDL.Continuous.LimPID
+      </code>
+      <a name="FromModelica.ExtendsClause.Buildings.Controls.OBC.CDL.Continuous.LimPID"></a></td>
+      <td>model with extends clause</td>
```
This is parsed incorrectly: extending a model is different from instantiating a model. The way this is documented is as if it were to instantiate `LimPID`.

JayHuLBL commented:
@mwetter Ready for your review. Following items have been implemented: 
- [x] updated parser to parse extends statement
- [x] updated arguments so to have `cdl` and `modelica` parsing mode. In `cdl` mode, it parses single .mo file; in `modelica` mode, it parses a package. 
- [x] re-organized output files structure according to the parsing mode. For example, in `cdl` mode, it has only one `html` for each model that includes all instantiated classes in it; in `modelica` mode, it has only one `html` for one package.
- [x] regenerated reference outputs for testing, updated `test` folder and `test\test_parser.js`

mwetter commented on lib/htmlWriter.js:
```diff
@@ -416,38 +495,43 @@ function getLibraryLink (className) {
-      setParameterHTMLAttributes(par)
-      setHtmlToParameterModification(par)
-    })
+  for (var i = 0; i < components.length; i++) {
```
i is not used in this loop

mwetter commented on app.js:
```diff
@@ -10,15 +10,8 @@ const ArgumentParser = require('argparse').ArgumentParser
   version: '0.0.1',
   addHelp: true,
-  description: 'CDL parser'
+  description: 'Buildings modelica parser'
```
Remove `Buildings`

mwetter commented on Makefile:
```diff
@@ -41,11 +41,21 @@ test-moParser:
+		node ../../app.js -l warn -f $${ff} -o html -m cdl; \
 		done)
-	rm -f test/FromModelica/modelica-json.log
+	cp -r test/FromModelica/json test/FromModelica/cdl && rm -R test/FromModelica/json
```
Introduce a flag `-d|--directory` to specify the output directory, with the default being the current directory.

mwetter commented on test/FromModelica/cdl/simplified/FromModelica.Block1.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "modelicaFile": "./Block1.mo",
```
Remove the `./`

mwetter commented on test/FromModelica/modelica/json/FromModelica.Block1.json:
```diff
@@ -0,0 +1,22 @@
+          "long_class_specifier": {
+            "name": "Block1",
+            "comment": "\"A block that instantiates nothing\"",
+            "composition": {
```
Remove empty objects

mwetter commented on test/FromModelica/modelica/simplified/FromModelica.Parameter1WithVendorAnnotation.json:
```diff
@@ -1,6 +1,6 @@
 [
   {
-    "modelicaFile": "./Parameter1WithVendorAnnotation.mo",
+    "modelicaFile": "FromModelica/Parameter1WithVendorAnnotation.mo",
```
Make sure the vendor annotations are included.

JayHuLBL commented:
This PR can be closed, as the `extends` statement will be parsed correctly after the new json output structure changed in branch issue157.
